[Profiling] Add upgrade docs for 8.9 (#2995)
elastic/observability-docs: docs/en/observability/profiling-upgrade.asciidoc

Universal Profiling is a beta program, and we are developing and iterating over it frequently upon receiving feedback.
For this reason, you have to erase *all existing profiling data* from your deployment when upgrading between minor releases in the {stack}.

> If you are upgrading from 8.8 we also provide a script that allows you to keep your data. When the product hits general availability,

Reviewer: ```suggestion
When upgrading from version 8.8, you can use the migration script to keep your data. When the product hits general availability,

```

Universal Profiling is a beta program, and we are developing and iterating over it frequently upon receiving feedback.
For this reason, you have to erase *all existing profiling data* from your deployment when upgrading between minor releases in the {stack}.
When upgrading from version 8.8, you can use the migration script to keep your data. When the product hits general availability,

> 

Reviewer: ```suggestion

```

. Perform a stack upgrade in {ecloud}
. Stop profiling data ingestion

> . When upgrading from version 8.8.0 or better: Migrate existing profiling data

Reviewer: ```suggestion
. *When upgrading from version 8.8.0 or more recent*: Migrate existing profiling data

```

> . When upgrading from versions before 8.8.0: Delete existing profiling data and setup Universal Profiling from scratch

Reviewer: ```suggestion
. *When upgrading from versions prior to 8.8.0*: Delete existing profiling data and setup Universal Profiling from scratch

```

. Perform a stack upgrade in {ecloud}
. Stop profiling data ingestion

> . *When upgrading from version 8.8.0 or more recent*: Migrate existing profiling data

Reviewer 2: I have a doubt on this one: previously (up until 8.8) we never created the `profiling-metrics` data stream in customers deployment.
When running 8.9 and the new Elasticsearch-based index management, will the data stream be created?

Author: Yes, the plugin will create any indices that are missing. Existing component and composable templates will be upgraded to their latest version and the ILM policy has an entirely name (we'll phase out the old one eventually).

. *When upgrading from versions prior to 8.8.0*: Delete existing profiling data and setup Universal Profiling from scratch

> 

Reviewer 2: I know it's not part of this PR, but I spotted a potential problem a little further down.
When we say `Perform a stack upgrade`, I think we should include a step that includes `Stop profiling data ingestion` between point 3 and 4.

Author: I can swap the steps, so it says:

1. Stop profiling data ingestion
2. Perform a stack upgrade in Elastic Cloud
3. When upgrading from version 8.8.0 or more recent: Migrate existing profiling data
4. When upgrading from versions prior to 8.8.0: Delete existing profiling data and setup Universal Profiling from scratch
5. Start profiling data ingestion

I'll also reorder the sections accordingly.

. Perform a stack upgrade in {ecloud}
. Stop profiling data ingestion
. *When upgrading from version 8.8.0 or more recent*: Migrate existing profiling data

> 

Reviewer: ```suggestion

```

NOTE: When stopping incoming requests, host-agent replicas back off and retry connecting to the {ecloud} endpoint at increasing time intervals.

[discrete]
[[profiling-migrate-data]]
=== Migrate existing profiling data

> For upgrades to 8.9 we offer a migration script if you meet the following requirements:

Reviewer: ```suggestion
When upgrading to version 8.9, you can use a migration script to migrate existing data if you meet the following requirements:

```

* The cluster has been upgraded from 8.8 to 8.9.
* The migration is performed on a Linux workstation with `jq` and `curl` installed.

> If you meet these requirements you may run the provided script below. Otherwise we suggest to delete existing profiling data and set up Universal Profiling from scratch.

Reviewer: ```suggestion
If you meet these requirements, follow the steps below to run the migration script. If not, we suggest <<profiling-delete-data, deleting your existing profiling data>> and <<profiling-from-scratch, setting up Universal Profiling from scratch>>.

```

> We also recommend to verify that you have a recent snapshot of your data so you can restore your data in case there is an error during the migration.

Reviewer: ```suggestion
You should also verify that you have a recent snapshot of your data so you can restore your it if there's an error during the migration.

```

    }
  }'

  echo "Reenabling ILM..."
  curl -s -f -o /dev/null -X POST -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_ilm/start"

  printf "Checking cluster health ... "
  local status=$(curl -s -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cluster/health?wait_for_status=green&wait_for_no_initializing_shards=true&wait_for_no_relocating_shards=true" | jq -r ".status")
  if [ "$status" == "red" ]; then
    printf "[ERROR: status is red]\n"
  else
    printf "[OK: status is %s]\n" "$status"
  fi

  {
    echo "Indices and aliases after migration"
    curl -s -f -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cat/indices/*profiling*?v&s=index&expand_wildcards=all"
    curl -s -f -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cat/aliases/*profiling*?v&expand_wildcards=all"
  } >> "elastic_universal_profiling_migration_log.txt"
}

function main {
  check_preconditions
  ask_user_prerequisites
  sanity_check_cluster
  migrate

  echo "Migration has finished. Please perform the following tasks:"
  echo ""
  echo "* Start profiling data ingestion"
  echo "* Enable Elastic Universal Profiling in Kibana"
}

main
----
====

[discrete]
[[profiling-delete-data]]
=== Delete existing profiling data

> _This step is required for upgrades from versions before 8.8.0. For upgrades from 8.8.0 or better you can <<profiling-migrate-data,migrate your data>>._

Reviewer: ```suggestion
NOTE: This step is required for upgrades from versions older than 8.8.0. For upgrades from 8.8.0 or newer, you can <<profiling-migrate-data,migrate your data>>.

```

NOTE: When stopping incoming requests, host-agent replicas back off and retry connecting to the {ecloud} endpoint at increasing time intervals.

[discrete]
[[profiling-migrate-data]]
=== Migrate existing profiling data

When upgrading to version 8.9, you can use a migration script to migrate existing data if you meet the following requirements:

* The cluster has been upgraded from 8.8 to 8.9.
* The migration is performed on a Linux workstation with `jq` and `curl` installed.

> If you meet these requirements, follow the steps below to run the migration script. If not, we suggest <<profiling-delete-data, deleting your existing profiling data>> and <<profiling-from-scratch, setting up Universal Profiling from scratch>>.

Reviewer 3: At this point I went looking through the prior text to find a link to the script for about 30 seconds before realizing that it follows below -- might just be me though.

Author: Good point. How about we reword this sentence to:

> If you meet these requirements, follow the steps below to migrate your data.

This just mentions the goal ("migrate the data") and directs the user to the following steps. The solution (i.e. the script) is only mentioned in the steps and at that point it's hopefully easier to spot. One note: I would have wished we could just provide a download link instead of rendering the script inline but the inline approach seems to be the established practice.

Reviewer 3: Sounds good!

You should also verify that you have a recent snapshot of your data so you can restore it if there's an error during the migration.

To run the migration script:

. Save the script below on a Linux workstation, e.g. as `migrate-8.9.sh`
. Make it executable: `chmod u+x migrate-8.9.sh`
. Ensure that `jq` and `curl` are installed on the Linux workstation.
. Export required environment variables:
  *  `export ES_USER=<superuser name>`
  *  `export ES_PASSWORD=<superuser password>`
  *  `export ES_HOST=<elasticsearch endpoint URL>`
. Run the migration: `./migrate-8.9.sh`
. Unset environment variables again: `unset ES_USER ES_PASSWORD ES_HOST`

[%collapsible]
.Migration Script
====
[source,bash]
----

> #!/usr/bin/env bash

Reviewer 3: Marking all of this text and copying it resulted in 2 blank lines to be copied before the shebang, which then resulted in this script being executed by `sh` instead of `bash` on my box where it bailed out at `set -o pipefail`. Not sure if we can do much about this though.

Author: That's not great and I could reproduce this. 

I played around a bit with CSS styles on the preview page but couldn't find the culprit. The only thing I noticed is that the behavior is identical to the ES docs. E.g. when you copy the snippet in the [index API docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-index_.html#create-document-ids-automatically) you'll also notice the two leading and trailing lines.

@mdbirnstiehl is there a possibility to render a "Copy to clipboard" button at the bottom of the script block (similar to the ES docs) to simplify Copy & Paste for users?

Reviewer 3: As a pretty terrible workaround we could add an `if [ $0 == "sh" ]` and then re-run with bash if necessary. But yeah, fixing the CSS (or whatever is causing this) would probably improve things in general (for all other snippets as well). I also recently noticed empty lines to be copied when following instructions for creating an Elastic setup in K8S and it was rather annoying.

Author: Thanks for the workaround. I'll keep this in mind and will add this if we don't find another way to address this (i.e. fixing the CSS).

Reviewer: I believe the only code block source that adds the copy button is [source,console]. If that is changed, you'll see the copy button as shown below. But that will impact how the script renders in the UI. Not sure if that is a big deal or not in this instance.

![image](https://github.com/elastic/observability-docs/assets/114418652/68e70657-76f6-4d73-a49c-5a6a3c700bd3)

Author: Thanks Mike for the clarification. In that case we'll probably go with the workaround that Joel has proposed in https://github.com/elastic/observability-docs/pull/2995#discussion_r1246477490.

Author: I played around with this a bit more and it turns out that 

* different OS configurations behave differently. For example, `/usr/bin/sh` is simply symlinked to `/usr/bin/bash`.
* it's not straightforward to determine in which shell you're actually running (see [whatshell.sh](https://www.in-ulm.de/~mascheck/various/whatshell/whatshell.sh.comments.html) for a rather complete implementation). 

I figured the safest option is to ask the user to run the script in bash by changing the instructions from:

```
./migrate-8.9.sh
```

to

```
bash migrate-8.9.sh
```

This ensures that the command is run in bash regardless whether the shebang is detected.

  {
    "is_write_index": true
  }'
}

function migrate {
  {
    echo "Indices and aliases before migration"
    curl -s -f -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cat/indices/*profiling*?v&s=index&expand_wildcards=all"
    curl -s -f -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cat/aliases/*profiling*?v&expand_wildcards=all"
  }  > "elastic_universal_profiling_migration_log.txt"

  echo "Allowing wildcard deletes ..."
  curl -s -f -o /dev/null -X PUT -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cluster/settings" -H 'Content-Type: application/json' -d'
  {
    "persistent": {
      "action.destructive_requires_name": false
    }
  }'

  echo "Disabling ILM temporarily..."
  curl -s -f -o /dev/null -X POST -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_ilm/stop"

  echo "Deleting obsolete indices..."
  curl -s -f -o /dev/null -X DELETE -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/.profiling-ilm-lock"
  curl -s -f -o /dev/null -X DELETE -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/profiling-returnpads-private,profiling-sq-executables,profiling-sq-leafframes,profiling-symbols?ignore_unavailable=true"

>   # We rely only on the "next" indices

Author: As an explicit note: In the following lines we delete the "current" indices. We can't keep both easily (a custom reindex procedure would be possible but IMHO way too complex) so we only use the "next" indices which will contain data since the last rollover but not before. So customers may see partial data for older profiles.

  fi

  echo "Renaming existing fields..."
  curl -s -f -o /dev/null -X PUT -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/profiling-events-*/_mapping" -H 'Content-Type: application/json' -d'
  {
    "properties": {
      "profiling.project.id": {
        "type": "alias",
        "path": "service.name"
      }
    }
  }'

  echo "Enabling Universal Profiling index template management and disabling wildcard deletes..."
  curl -s -f -o /dev/null -X PUT -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cluster/settings" -H 'Content-Type: application/json' -d'
  {
    "persistent": {
      "xpack.profiling.templates.enabled": true,
      "action.destructive_requires_name": true
    }
  }'

  echo "Reenabling ILM..."
  curl -s -f -o /dev/null -X POST -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_ilm/start"

  printf "Checking cluster health ... "
  local status=$(curl -s -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cluster/health?wait_for_status=green&wait_for_no_initializing_shards=true&wait_for_no_relocating_shards=true" | jq -r ".status")
  if [ "$status" == "red" ]; then
    printf "[ERROR: status is red]\n"
  else
    printf "[OK: status is %s]\n" "$status"

>   fi

Reviewer 3: I tend to deploy my testing clusters without ES replication to save us some money & this check prevents me from upgrading such deployments. Should we allow yellow? Not sure. Perhaps alternatively we could add an argument to allow skipping this?

Author: Actually we do allow a `yellow` status. The script output should have been

> [OK: status is yellow]

And then the script should have continued normally. Did you see a different behavior?

Author: Ah, I was mislead by the snippet. There are actually two checks:

1. Before the migration to ensure the cluster is in a healthy state. Indeed, there we check for cluster health `green`
2. After the migration to make sure nothing is off. We are more lenient here and allow `yellow` because it's possible that replicas are not fully up yet after the migration (they will be eventually)

I think we can be probably a bit more lenient and allow cluster health `yellow` in the first check. I've pushed the change in https://github.com/elastic/observability-docs/pull/2995/commits/14f3e39f71edb2c26b1e35c45485d35c9f58bfd5.

Reviewer 3: Ah, sorry for the confusion with quoting the wrong snippet.

  {
    echo "Indices and aliases after migration"
    curl -s -f -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cat/indices/*profiling*?v&s=index&expand_wildcards=all"
    curl -s -f -k -u "$ES_USER":"$ES_PASSWORD" "$ES_HOST/_cat/aliases/*profiling*?v&expand_wildcards=all"
  } >> "elastic_universal_profiling_migration_log.txt"
}

function main {
  check_preconditions
  ask_user_prerequisites
  sanity_check_cluster
  migrate

  echo "Migration has finished. Please perform the following tasks:"
  echo ""
  echo "* Start profiling data ingestion"

>   echo "* Enable Elastic Universal Profiling in Kibana"

Reviewer 2: Do you think we can remove this line? Profiling is enabled by default in all Cloud deployments since 8.9.

Author: Good catch. I've reworded the entire statement now to:

> Migration has finished. You can now start profiling data ingestion.

NOTE: When stopping incoming requests, host-agent replicas back off and retry connecting to the {ecloud} endpoint at increasing time intervals.

[discrete]
[[profiling-migrate-data]]
=== Migrate existing profiling data

When upgrading to version 8.9, you can use a migration script to migrate existing data if you meet the following requirements:

* The cluster has been upgraded from 8.8 to 8.9.
* The migration is performed on a Linux workstation with `jq` and `curl` installed.

If you meet these requirements, follow the steps below to run the migration script. If not, we suggest <<profiling-delete-data, deleting your existing profiling data>> and <<profiling-from-scratch, setting up Universal Profiling from scratch>>.

> You should also verify that you have a recent snapshot of your data so you can restore your it if there's an error during the migration.

Reviewer: ```suggestion
You should also verify that you have a recent snapshot of your data so you can restore it if there's an error during the migration.

```

> 

Reviewer: ```suggestion

```
